ROX-16426 / ROX-14418: Get ENP up and running

Maddosaurus opened the pull request:
~~FIXME: After review, before merging, remove the `dry-run` flag from the production `run-and-upload` step [here](https://github.com/stackrox/external-network-pusher/pull/24/files#diff-920dae94a370229ec0d6613197b3d00e75ed7942250e00ff40d6414d48f61195R106).~~

This PR updates and extends the workflow.
It introduces two separate steps as follow ups to the build step:
- `run-dry-run`: Runs the binary in dry-run mode. This will automatically run on PRs and can be manually triggered on the "Actions" tab 
- `run-and-upload`: Runs the binary in production mode where it writes to the GCP bucket `defitions.stackrox.io`. This will be run daily on 10.30 AM UTC and can also be triggered manually on the "Actions" tab (by unchecking the dry run option)

Additionally, it will send a notification to a Team Maple Slack channel if the scheduled run failed.

vikin91 commented on .github/workflows/push-networks.yaml:
```diff
@@ -1,13 +1,24 @@
 name: Build external-network-pusher
 
 on:
+  pull_request:
+    types:
+    - opened
+    - reopened
+    - synchronize
```
Why do we want to run this on PRs?

vikin91 replied:
Ups, I see now, that we are not in Stackrox repo 🤦  Sorry, please ignore this comment :)

Maddosaurus replied:
The idea is to build, test, and do a dry-run on PRs against this repo to ensure that the resulting binary will succeed 😄 
We're not in our main repo, so this should be fine, as it's the only automated check that is running.

vikin91 commented on .github/workflows/push-networks.yaml:
```diff
@@ -1,13 +1,24 @@
 name: Build external-network-pusher
```
I know you did not change it, but is this action really building it? I think that calling it `Run external-...` would make it clear. Or better: `Update external network definitions on definitions.stackrox.io`

Maddosaurus replied:
I did introduce the name, but it was in an earlier PR. I agree that `build` is not exactly accurate.
I like `Update external network definitions` better, let me think about it and see what I can come up with.

vikin91 commented on .github/workflows/push-networks.yaml:
```diff
@@ -60,7 +103,5 @@ jobs:
       
       - name: Run external-network-pusher
-        run:  linux/network-crawler --dry-run --bucket-name 123
+        run:  linux/network-crawler --dry-run --bucket-name definitions.stackrox.io  # FIXME: Remove dry-run flag here
```
Is the resulting file that would be pushed viewable/readable for the human? Could we maybe print it to the action output, so that we know what go written?

Other ideas: Could we download the previous file from definitions.stackrox.io before overwriting it and print it too? Maybe with some diff as well. This would help us understand in the future what this tool is actually doing.

(If the bucket is versioned, then the suggestions from this comment have less value, but I wanted to ask anyway)

vikin91 reviewed (commented):
It looks really good! 
I have some questions and optional suggestions.

vikin91 reviewed (approved):
LGTM, but please let Tom give it a quick look before merging.

tommartensen commented on .github/workflows/push-networks.yaml:
```diff
@@ -44,12 +55,44 @@ jobs:
+          credentials_json: '${{ secrets.GCP_NETWORKS_UPLOADER_SA }}'
+      - name: 'Set up Cloud SDK'
+        uses: 'google-github-actions/setup-gcloud@v0'
+        with:
+          install_components: "gke-gcloud-auth-plugin"
```
```suggestion
```

This is only required if you want to authenticate to a GKE cluster, you can remove these two lines.

tommartensen commented on .github/workflows/push-networks.yaml:
```diff
@@ -44,12 +55,44 @@ jobs:
+          credentials_json: '${{ secrets.GCP_NETWORKS_UPLOADER_SA }}'
+      - name: 'Set up Cloud SDK'
+        uses: 'google-github-actions/setup-gcloud@v0'
+        with:
+          install_components: "gke-gcloud-auth-plugin"
```
```suggestion
```

See above.

tommartensen commented on .github/workflows/push-networks.yaml:
```diff
@@ -44,12 +55,44 @@ jobs:
+      
+      - name: Dry run external-network-pusher
+        run:  linux/network-crawler --dry-run --bucket-name definitions.stackrox.io
+        continue-on-error: false
```
What is the reason for `continue-on-error` ?

Maddosaurus replied:
This was a carry over from the kubelinter workflow I based this one on.
From what I can see, this is set to `false` by default anyways, so I might as well remove it from here.

tommartensen reviewed (changes requested):
Two suggestions and a question.

One more thing to consider: How do you want to monitor job success for the scheduled runs?

tommartensen commented:
> There are mail notifications that are being sent out if a job fails, but maybe you have a different mechanism in mind?
My gut feeling tells me something like a Slack bot posting to acs-nightly-ci-runs or acs-test-failures might also not be a bad idea. WDYT?

From https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/notifications-for-workflow-runs: 

> Notifications for scheduled workflows are sent to the user who initially created the workflow. If a different user updates the cron syntax in the workflow file, subsequent notifications will be sent to that user instead. If a scheduled workflow is disabled and then re-enabled, notifications will be sent to the user who re-enabled the workflow rather than the user who last modified the cron syntax.

Mail notifications to a user (who might or might not read their emails, leave the organization, ...) sounds like a single point of failure to me. I'll leave the decision to you and give you a LGTM already.

tommartensen reviewed (approved):
LGTM

Maddosaurus commented:
@tommartensen I have added Slack notifications for a team Maple channel. Please feel free to take another look [here](https://github.com/stackrox/external-network-pusher/pull/24/files#diff-920dae94a370229ec0d6613197b3d00e75ed7942250e00ff40d6414d48f61195R120).

@vikin91 I've renamed the workflow and added a [download of the current networks and checksum](https://github.com/stackrox/external-network-pusher/pull/24/files#diff-920dae94a370229ec0d6613197b3d00e75ed7942250e00ff40d6414d48f61195R21) before overwriting them. They will show up in the artifacts tab of a run and will be cached for 60 days. [Example here](https://github.com/stackrox/external-network-pusher/actions/runs/4676819671) (see "Artifacts" at the bottom).

tommartensen commented:
@Maddosaurus - Slack notification LGTM
